pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -374,6 +391,78 @@ func (uh *upgradeHandler) toUpgradeCluster(cluster *v3.Cluster) (bool, error) {
 	return false, nil
 }
 
+func (uh *upgradeHandler) restore(cluster *v3.Cluster) error {
+	nodes, err := uh.nodeLister.List(cluster.Name, labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errgrp errgroup.Group
+	nodesQueue := util.GetObjectQueue(nodes)
+	for w := 0; w < 5; w++ {
+		errgrp.Go(func() error {
+			var errList []error
+			for node := range nodesQueue {
+				node := node.(*v3.Node)
+				if node.Status.NodeConfig != nil && workerOnly(node.Status.NodeConfig.Role) {
+					if node.Status.NodePlan.Version != cluster.Status.NodeVersion {
+						if err := uh.setNodePlan(node, cluster, false); err != nil {
+							errList = append(errList, err)
+						}
+						logrus.Infof("cluster [%s]: updated node [%s] for restore, plan version [%v]", cluster.Name,
+							node.Name, cluster.Status.NodeVersion)
+					}
+				}
+			}
+			return util.ErrList(errList)
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+
+	clusterCopy := cluster.DeepCopy()
+	clusterCopy.Annotations[clusterprovisioner.RkeRestoreAnnotation] = "false"
+	if _, err := uh.clusters.Update(clusterCopy); err != nil {
+		if !errors.IsConflict(err) {
+			return err
+		}
+		return uh.retryClusterUpdate(cluster.Name)
+	}
+	return nil
+}
+
+func (uh *upgradeHandler) retryClusterUpdate(name string) error {
+	backoff := wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1,
+		Jitter:   0,
+		Steps:    7,
+	}
+
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		cluster, err := uh.clusters.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if cluster.Annotations[clusterprovisioner.RkeRestoreAnnotation] != "false" {
+			clusterCopy := cluster.DeepCopy()
```

Comment:
This may perform a deep copy in a loop. 
This object has already been copied once on https://github.com/rancher/rancher/pull/26032/files#diff-a16c43ed026936b620e2fdb9ca6196b8R424 (line 424)